### PR TITLE
Use fileid as string in web UI

### LINKED ISSUE #30434: Devil in the details: File ID javascript limit
@DeepDiver1975 @felixboehm @butonic @PVince81 

We have hit an issue that may impact other big ownCloud deployments.

ownCloud relies on a file id to be an integer, currently the size has been increased server-side to unsigned 64 bits: https://github.com/owncloud/core/issues/26901

The problem is that Javascript only support numerical precision up to 2^53.

Long story short: the fileInfo model in JS relies on the fileId to be a number and it calls parseInt("file id") but if the number is bigger than 2^53 results in an erratic conversion, see image:

the-devil-in-the-details

Result: every server side application  relying on the file Id sent from the browser will fail as the file id is not correct and can point to an non-existing file or to another file in other user home directory.

The solution for this problem is to make the file ID a string and not a number, the sync client took this sage decision long time ago.

Cheers,
Hugo

### PR DESCRIPTION
**Pull request by PVince81**
## Description
Because of limitations to JS max int, we now use strings instead of
integers.

## Related Issue
Fixes https://github.com/owncloud/core/issues/30434

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by PVince81** [comment]
- [ ] TODO: fix sharing "item_source" and "file_source"
- [ ] TODO: federated share "file_id" field

**Comment by PVince81** [comment]
- [ ] versions file id

**Comment by PVince81** [comment]
- [ ] activity file id...

**Comment by PVince81** [comment]
- [ ] previews in activity seem broken

**Comment by PVince81** [comment]
- [ ] gallery slideshow broken...

**Comment by PVince81** [comment]
seems previews and gallery already broken on master, so likely unrelated to this change

**Comment by PVince81** [comment]
I think I've nailed most of the cases.

**Comment by phil-davis** [comment]
First run of drone had UI test fails at:
```
sharingInternalGroupsUsers
    And the sharing permissions of "User One" for "simple-folder" are set to # SharingContext::theSharingPermissionsOfAreSetTo()
| delete | no |

sharingExternalOther
    And the sharing permissions of "user1@%remote_server% (federated)" for "simple-folder" are set to # SharingContext::theSharingPermissionsOfAreSetTo()
| delete | no |

restrictSharing
    And the sharing permissions of "User One" for "simple-folder" are set to # SharingContext::theSharingPermissionsOfAreSetTo()
| share | no |
```
these are the only 3 places that the step ``the sharing permissions of "user" for "folder" are set to`` is used.

So I suspect that there is some issue with the sharing permissions checkboxes. I don't see immediately in the UI test code where it tries to traverse any HTML elements... that contain the file id in their ``xpath``.

**Comment by PVince81** [comment]
@phil-davis seems those problems disappeared?

**Comment by phil-davis** [comment]
@PVince81 yes, after you pushed more changes it was passing, so I don't know if the first fails were just unfortunate (it seemed a big coincidence that it was sharing test steps that failed in the 3 sharing UI test suites).

**Comment by PVince81** [comment]
pinging more people to review the JS changes

File: apps/files_sharing/lib/API/Remote.php [hunk 1]
```diff
@@ -134,7 +134,7 @@ private static function extendShareInfo($share) {
 		$share['mtime'] = $info->getMtime();
 		$share['permissions'] = $info->getPermissions();
 		$share['type'] = $info->getType();
-		$share['file_id'] = $info->getId();
+		$share['file_id'] = '' . $info->getId();
```
Comment (VicDeo) [inline apps/files_sharing/lib/API/Remote.php]: http://php.net/manual/en/function.strval.php ?

**Review by VicDeo — COMMENTED** [review]
The only question is about the way of casting to string

File: apps/files_sharing/lib/API/Remote.php (hunk 1 above)
Comment (PVince81) [reply apps/files_sharing/lib/API/Remote.php]: meh

**Comment by PVince81** [comment]
anything else or does the code look ok ?

**Review by VicDeo — APPROVED** [review]
@PVince81 I'd vote against hardcore casting by concatenation.
there are more clean ways to get string in PHP: `strval(5)` or `(string) 5`

But it's up to you.
No other objections.

**Comment by PVince81** [comment]
@VicDeo ok, since you insisted I'll replace "hardcore casting" 😄

**Comment by PVince81** [comment]
replaced 4 occurrences to strval in PHP code and rebased + squashed

**Comment by PVince81** [comment]
why is this upload test failing ? it seems to not find the lorem file

**Comment by phil-davis** [comment]
It is usually related to a lock left somewhere - I think the upload-overwrite fails because the file is already locked, then the content is checked and of course is the old content, not the expected new content.

If you get unlucky, this happens to you. This is the one I notice happening the most.

**Comment by phil-davis** [comment]
Go to the owncloud-log step and search for ``locked`` - if that can be fixed then other troubles will go away.

**Comment by phil-davis** [comment]
```
{"reqId":"tw5k1PGffGkswoDsxK13","level":0,"time":"2018-02-16T10:27:06+00:00","remoteAddr":"172.18.0.3","user":"user1","app":"core","method":"GET","url":"\/remote.php\/dav\/files\/user1\/simple-folder\/lorem.txt?c=1478997b23797c954f7ee8ea250c4c99&x=32&y=32&forceIcon=0&preview=1","message":"Generating preview for \"\/user1\/files\/simple-folder\/lorem.txt\" with \"OC\\Preview\\TXT\""}
{"reqId":"evWmtS6qlZ3OLkHVeCOM","level":4,"time":"2018-02-16T10:27:07+00:00","remoteAddr":"172.18.0.3","user":"user1","app":"webdav","method":"PUT","url":"\/remote.php\/dav\/files\/user1\/simple-folder\/lorem.txt","message":"Exception: {\"Message\":\"HTTP\\\/1.1 423 \\\"simple-folder\\\/lorem.txt\\\" is locked\",\"Exception\":\"OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Exception\\\\FileLocked\",\"Code\":0,\"Trace\":\"#0 \\\/drone\\\/src\\\/lib\\\/public\\\/Events\\\/EventEmitterTrait.php(50): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\File->OCA\\\\DAV\\\\Connector\\\\Sabre\\\\{closure}()\\n#1 \\\/drone\\\/src\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/File.php(258): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\File->emittingCall(Object(Closure), Array, 'file', 'create')\\n#2 \\\/drone\\\/src\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(1129): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\File->put(Resource id #53)\\n#3 \\\/drone\\\/src\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(513): Sabre\\\\DAV\\\\Server->updateFile('files\\\/user1\\\/sim...', Resource id #53, NULL)\\n#4 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpPut(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#5 \\\/drone\\\/src\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#6 \\\/drone\\\/src\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:PUT', Array)\\n#7 \\\/drone\\\/src\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#8 \\\/drone\\\/src\\\/apps\\\/dav\\\/lib\\\/Server.php(279): Sabre\\\\DAV\\\\Server->exec()\\n#9 \\\/drone\\\/src\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#10 \\\/drone\\\/src\\\/remote.php(175): require_once('\\\/drone\\\/src\\\/apps...')\\n#11 {main}\",\"File\":\"\\\/drone\\\/src\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/File.php\",\"Line\":206,\"User\":\"user1\"}"}
```
suspiciously related to a "generating preview" message above it.

**Comment by PVince81** [comment]
raised here: https://github.com/owncloud/core/issues/30506

**Comment by PVince81** [comment]
trying my luck again at the CI roulette...

**Comment by phil-davis** [comment]
@PVince81  you will win if you rebase

**Comment by PVince81** [comment]
@ownclouders rebase

**Comment by ownclouders** [comment]
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

**Comment by ownclouders** [comment]
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

**Comment by phil-davis** [comment]
You are so lucky
https://drone.owncloud.com/owncloud/core/4034/137
```
+ git init
Initialized empty Git repository in /drone/src/.git/
+ git remote add origin https://github.com/owncloud/core.git
+ git fetch --no-tags --depth=50 origin +refs/pull/30435/merge:
fatal: The remote end hung up unexpectedly
exit status 128 
```
I restarted ``drone``

**Comment by PVince81** [comment]
thanks for the luck. maybe someone should start a service where people can pool their luck in a single location for greater luck...

**Comment by phil-davis** [comment]
@PVince81 Is this to be backported so that these huge file ids can work in a 10.0.* release?

**Comment by PVince81** [comment]
This is a bit dangerous, so no

**Comment by labkode** [comment]
@PVince81 We will need to have this functionality for our service, do you expect to to release for 10.1.*?

**Comment by PVince81** [comment]
This will be in the next major release, yes. But not the minor one due to bigger risks.

I'm not sure whether the current automated tests already cover for all the cases touched by this PR.
If not, it would require to schedule a regression test of the complete files app and sharing UI.

**Pull request merged**